File: JavaScript/Array&String/238. Product of Array Except Self.js
/* 238. Product of Array Except Self 
Difficulty: Medium
Submission Link: https://leetcode.com/problems/product-of-array-except-self/submissions/1030277210/
*/
var productExceptSelf = function (nums) {

    let result = [];
    let left = 1;
    let right = 1;
  
    for (let i = 0; i < nums.length; i++) {
      result[i] = left;
      left *= nums[i];
    }
  
    for (let i = nums.length - 1; i >= 0; i--) {
      result[i] *= right;
      right *= nums[i];
    }
  
    return result;
  };

  module.exports = productExceptSelf;

File: Python3/Array&String/238. Product of Array Except Self.py
""" 238. Product of Array Except Self
Difficulty: Medium
Submission Link: https://leetcode.com/problems/product-of-array-except-self/submissions/1030276923/ """

class Solution:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        res = [1]*len(nums)

        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix *= nums[i]

        postfix = 1
        for i in range(len(nums)-1,-1,-1):
            res[i] *= postfix
            postfix *= nums[i]
        
        return res